support/metrics/metricsTracker.go

```diff
@@ -0,0 +1,165 @@
+	client *http.Client
+	props  props
+	start  time.Time
+}
+
+type event struct {
+	UserID    string      `json:"user_id"`
+	DeviceID  string      `json:"device_id"`
+	EventType string      `json:"event_type"`
+	Props     interface{} `json:"event_properties"`
+}
+
+// props holds the properties that we need for all Amplitude events.
+// This lives on the `Tracker` struct, although
+// TODO: Add geodata.
+// TODO: Add cloud server information.
+type props struct {
+	BuildVersion string    `json:"build_version"`
+	Os           string    `json:"os"`
+	Gui          bool      `json:"gui"`
+	Strategy     string    `json:"strategy"`
+	UpdateTime   float64   `json:"update_time"`
+	Exchange     string    `json:"exchange"`
+	Pair         string    `json:"pair"`
+	SessionID    uuid.UUID `json:"session_id"`
+}
+
+// updateProps holds the properties for the update Amplitude event.
+type updateProps struct {
+	props
+	SecondsSinceStart float64 `json:"seconds_since_start"`
+}
+
+// deleteProps holds the properties for the delete Amplitude event.
+type deleteProps struct {
+	props
+	SecondsSinceStart float64 `json:"seconds_since_start"`
+	Exit              bool    `json:"exit"`
+	StackTrace        string  `json:"stack_trace"`
+}
+
+// MakeMetricsTracker is a factory method to create a `metrics.Tracker`.
+func MakeMetricsTracker(
+	version string,
+	os string,
+	gui bool,
+	strategy string,
+	updateTime float64,
+	exchange string,
+	pair string,
+) *Tracker {
+	client := &http.Client{}
+	sessionID := mustSessionID()
+	props := props{
+		BuildVersion: version,
+		Os:           os,
+		Gui:          gui,
+		Strategy:     strategy,
+		UpdateTime:   updateTime,
+		Exchange:     exchange,
+		Pair:         pair,
+		SessionID:    sessionID,
+	}
+
+	t := Tracker{
+		client: client,
+		props:  props,
+		start:  time.Now(),
+	}
+	return &t
+}
+
+func mustSessionID() uuid.UUID {
+	sessionID, e := uuid.NewRandom()
+	if e != nil {
+		return [16]byte{}
+	}
+
+	return sessionID
+}
+
+// SendStartupEvent sends the startup Amplitude event.
+func (t *Tracker) SendStartupEvent(l logger.Logger) error {
+	return t.sendEvent(l, "ce:test_startup", t.props)
+}
+
+// SendUpdateEvent sends the update Amplitude event.
+func (t *Tracker) SendUpdateEvent(l logger.Logger) error {
+	props := updateProps{
+		props:             t.props,
+		SecondsSinceStart: time.Now().Sub(t.start).Seconds(),
+	}
+
+	return t.sendEvent(l, "ce:test_update", props)
+}
+
+// SendDeleteEvent sends the delete Amplitude event.
+func (t *Tracker) SendDeleteEvent(l logger.Logger, exit bool) error {
+	props := deleteProps{
+		props:             t.props,
+		SecondsSinceStart: time.Now().Sub(t.start).Seconds(),
+		Exit:              exit,
+		StackTrace:        "", // TODO: Determine how to do this.
+	}
+
+	return t.sendEvent(l, "ce:test_delete", props)
+}
+
+func (t *Tracker) sendEvent(l logger.Logger, eventType string, eventProps interface{}) error {
+	requestBody, e := json.Marshal(map[string]interface{}{
```

Comment:
Argh, I meant to ask this in my initial questions, my bad. That method only works for request body of type `map[string]string`, while the serializable struct contains fields with non-string values. 

The method for type `map[string]interface{}`, `JSONRequestDynamicHeaders` makes sense on its own. But it requires a map of `HeaderFn` defined per-field. I tried to understand how to create this map from usage, but this function is only used in `support/sdk/ccxt`. The usage there is entirely contextual to the package and basically impossible to model. I could not figure out how to generate a comparable map for my own use case.

How do I do this? Do you have a helpful link or tutorial?

Reply:
Ignore `JSONRequestDynamicHeaders()` for now. We need to use the `JSONRequest()` method. 

The `map[string]string` is for the `headers` variable.
It takes in the `data` field as a `string`. You will still need to marshall your data into a string like you've done here and then instead of using `http.Post` directly you should use `networking.JSONRequest()`

This function is pretty straightforward actually and not much different from `http.Post`. See callers of that function for usage examples. Most example use the `HTTP GET` but in our case we would use `"POST"` instead and instead of using an empty string for the `data` field it will be `string(requestBody)`, that's pretty much it, will make more sense once you find the usage examples in the codebase.